### PR TITLE
Allow waveform iteration for serial explicit coupling scheme

### PR DESCRIPTION
**Pull request by NiklasKotarsky**
## Main changes of this PR

Add support for waveform iterations for the serial-explicit coupling.

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [x] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I sticked to C++17 features.
* [x] I sticked to CMake version 3.16.3.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [x] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?

### CODE REVIEW

File: tests/serial/time/explicit/serial-coupling/ReadWriteScalarDataWithWaveform.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  double windowDt  = maxDt;
+  double dt        = windowDt / (nSubsteps - 0.5); // Solver always tries to do a timestep of fixed size.
+  double currentDt = dt > maxDt ? maxDt : dt;      // determine actual timestep length; must fit into remaining time in window
+  double timeCheckpoint;
+
+  while (precice.isCouplingOngoing()) {
+
+    precice.readScalarData(meshName, readDataName, vertexID, currentDt, readData);
+
+    if (context.isNamed("SolverOne")) { // in the first iteration of each window, we only have one sample of data. Therefore constant interpolation
```
Comment (BenjaminRodenberg) [inline tests/serial/time/explicit/serial-coupling/ReadWriteScalarDataWithWaveform.cpp:79]:
```suggestion
    if (context.isNamed("SolverOne")) { // first participant receives constant value from second
```

File: tests/serial/time/explicit/serial-coupling/ReadWriteScalarDataWithWaveform.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      BOOST_TEST(readData == readFunction(timeCheckpoint));
+    } else { // in the following iterations we have two samples of data. Therefore linear interpolation
```
Comment (BenjaminRodenberg) [inline tests/serial/time/explicit/serial-coupling/ReadWriteScalarDataWithWaveform.cpp:81]:
```suggestion
    } else { // second participant samples from waveform
```

File: tests/serial/time/explicit/serial-coupling/ReadWriteScalarDataWithWaveform.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      BOOST_TEST(readData == readFunction(time + currentDt));
+    }
+
+    precice.readScalarData(meshName, readDataName, vertexID, currentDt / 2, readData);
+
+    if (context.isNamed("SolverOne")) { // in the first iteration of each window, we only have one sample of data. Therefore constant interpolation
```
Comment (BenjaminRodenberg) [inline tests/serial/time/explicit/serial-coupling/ReadWriteScalarDataWithWaveform.cpp:87]:
```suggestion
    if (context.isNamed("SolverOne")) { // first participant receives constant value from second
```

File: tests/serial/time/explicit/serial-coupling/ReadWriteScalarDataWithWaveform.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+      BOOST_TEST(readData == readFunction(timeCheckpoint));
+    } else { // in the following iterations we have two samples of data. Therefore linear interpolation
```
Comment (BenjaminRodenberg) [inline tests/serial/time/explicit/serial-coupling/ReadWriteScalarDataWithWaveform.cpp:89]:
```suggestion
    } else { // second participant samples from waveform
```

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 5]
```diff
@@ -307,10 +307,29 @@ void CouplingSchemeConfiguration::xmlEndTagCallback(
   PRECICE_TRACE(tag.getFullName());
   if (tag.getNamespace() == TAG) {
     if (_config.type == VALUE_SERIAL_EXPLICIT) {
+
+      //Check the order of both participants in the explicit coupling
```
Comment (BenjaminRodenberg) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:311]:
```suggestion
      //Check the waveform order of both participants in the explicit coupling
```

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 6]
```diff
@@ -307,10 +307,29 @@ void CouplingSchemeConfiguration::xmlEndTagCallback(
...
+
+        auto first_participant = _participantConfig->getParticipant(first);
+        for (auto &dataContext : first_participant->readDataContexts()) {
+          int usedOrder = dataContext.getInterpolationOrder();
+          // The first participants waveform order has to be 0 for serial explicit coupling
+          if (usedOrder != 0) {
+            PRECICE_ERROR(
+                "You configured <read-data name=\"{}\" mesh=\"{}\" waveform-order=\"{}\" />, but for the serial explicit coupling scheme only a maximum waveform-order of \"{}\" is allowed for the first participant.",
+                dataContext.getDataName(), dataContext.getMeshName(), usedOrder, 0);
+          }
```
Comment (BenjaminRodenberg) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:324]:
```suggestion
          int allowedOrder = 0;
          if (usedOrder != allowedOrder) {
            PRECICE_ERROR(
                "You configured <read-data name=\"{}\" mesh=\"{}\" waveform-order=\"{}\" />, but for the serial explicit coupling scheme only a maximum waveform-order of \"{}\" is allowed for the first participant.",
                dataContext.getDataName(), dataContext.getMeshName(), usedOrder, allowedOrder);
          }
```

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 7]
```diff
@@ -307,10 +307,29 @@ void CouplingSchemeConfiguration::xmlEndTagCallback(
...
+        }
+        auto second_participant = _participantConfig->getParticipant(second);
+        for (auto &dataContext : second_participant->readDataContexts()) {
+          int usedOrder = dataContext.getInterpolationOrder();
+          PRECICE_ASSERT(usedOrder >= 0);
```
Comment (BenjaminRodenberg) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:329]: Why not also raise a proper error here? I think a variation of the one from above would be fine.

**Review by BenjaminRodenberg — CHANGES_REQUESTED** [review]

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp (hunk 7 above)
Comment (NiklasKotarsky) [reply src/cplscheme/config/CouplingSchemeConfiguration.cpp]: Yes, I  have added that now.

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 8]
```diff
@@ -307,10 +307,34 @@ void CouplingSchemeConfiguration::xmlEndTagCallback(
...
+                dataContext.getDataName(), dataContext.getMeshName(), usedOrder, allowedOrder);
+          }
+        }
+        auto second_participant = _participantConfig->getParticipant(second);
+        for (auto &dataContext : second_participant->readDataContexts()) {
+          int usedOrder = dataContext.getInterpolationOrder();
+          if (usedOrder < 0) {
+            PRECICE_ERROR(
+                "You configured <read-data name=\"{}\" mesh=\"{}\" waveform-order=\"{}\" />, but for the serial explicit coupling scheme a waveform-order greater than \"{}\" is allowed for the second participant.",
+                dataContext.getDataName(), dataContext.getMeshName(), usedOrder, 0);
```
Comment (BenjaminRodenberg) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:333]:
```suggestion
                "You configured <read-data name=\"{}\" mesh=\"{}\" waveform-order=\"{}\" />, but for the serial explicit coupling scheme the waveform-order must be non-negative for the second participant.",
                dataContext.getDataName(), dataContext.getMeshName(), usedOrder);
```

**Pull request merged**